File: backend/app/schemas/roleplay.py
"""Practice-arena (roleplay) schemas."""

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

# === Persona schemas ===


class ProspectPersonaResponse(BaseModel):
    """A synthetic prospect persona."""

    id: uuid.UUID
    workspace_id: uuid.UUID | None
    slug: str
    name: str
    description: str | None
    difficulty: str
    channel: str
    persona_prompt: str
    opening_message: str | None
    objections: list[str]
    goal: str | None
    is_builtin: bool
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ProspectPersonaCreate(BaseModel):
    """Create a custom workspace persona."""

    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    slug: str | None = Field(default=None, max_length=100)
    difficulty: str = "medium"
    channel: str = "sms"
    persona_prompt: str = Field(min_length=1)
    opening_message: str | None = None
    objections: list[str] = Field(default_factory=list)
    goal: str | None = None


class ProspectPersonaUpdate(BaseModel):
    """Update a custom workspace persona."""

    name: str | None = Field(default=None, max_length=200)
    description: str | None = None
    difficulty: str | None = None
    channel: str | None = None
    persona_prompt: str | None = None
    opening_message: str | None = None
    objections: list[str] | None = None
    goal: str | None = None


# === Rehearsal run schemas ===


class RehearsalTranscriptTurn(BaseModel):
    """One turn in a rehearsal transcript."""

    role: str
    content: str


class RehearsalRunSummary(BaseModel):
    """Lightweight rehearsal run for list views."""

    id: uuid.UUID
    workspace_id: uuid.UUID
    agent_id: uuid.UUID | None
    persona_id: uuid.UUID | None
    agent_name: str | None
    persona_name: str | None
    rehearsee: str
    channel: str
    status: str
    pending_action: str | None = None
    attempt_count: int = 0
    retryable: bool = False
    overall_score: float | None
    objection_coverage: float | None
    booking_attempted: bool | None
    tone_score: float | None
    created_at: datetime
    completed_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class RehearsalRunResponse(RehearsalRunSummary):
    """Full rehearsal run including transcript and report."""

    max_turns: int
    transcript: list[RehearsalTranscriptTurn]
    scores: dict[str, Any]
    strengths: list[str]
    gaps: list[str]
    suggestions: list[str]
    summary: str | None
    error: str | None
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class CreateRehearsalRequest(BaseModel):
    """Start a rehearsal of an agent (or human rep) against a persona."""

    idempotency_key: uuid.UUID
    agent_id: uuid.UUID
    persona_id: uuid.UUID
    rehearsee: Literal["ai", "human"] = "ai"
    channel: Literal["sms", "voice"] | None = None
    max_turns: int = Field(default=6, ge=1, le=12)


class HumanTurnRequest(BaseModel):
    """A human rep's reply during a live rehearsal."""

    message: str = Field(min_length=1, max_length=4000)
    expected_turn_count: int = Field(ge=1, le=25)


class RetryRehearsalRequest(BaseModel):
    """Retry only the failure the caller observed, not a later failed attempt."""

    expected_attempt_count: int = Field(ge=0)

File: frontend/src/types/roleplay.ts
// Practice-arena (roleplay) domain types.

export type PersonaDifficulty = "easy" | "medium" | "hard";

export interface ProspectPersona {
  id: string;
  workspace_id: string | null;
  slug: string;
  name: string;
  description: string | null;
  difficulty: string;
  channel: string;
  persona_prompt: string;
  opening_message: string | null;
  objections: string[];
  goal: string | null;
  is_builtin: boolean;
  created_at: string;
  updated_at: string;
}

export type RehearseeType = "ai" | "human";

export type RehearsalStatus = "pending" | "running" | "completed" | "failed";

export interface RehearsalTranscriptTurn {
  role: string; // "prospect" | "agent"
  content: string;
}

export interface ObjectionBreakdownItem {
  objection: string;
  addressed: boolean;
  note: string;
}

export interface RehearsalScores {
  overall_score?: number;
  objection_coverage_score?: number;
  tone_score?: number;
  tone_label?: string;
  booking_attempted?: boolean;
  objection_breakdown?: ObjectionBreakdownItem[];
  sentiment?: string | null;
  sentiment_score?: number | null;
  intents?: string[];
  topics?: string[];
}

export interface RehearsalRunSummary {
  id: string;
  workspace_id: string;
  agent_id: string | null;
  persona_id: string | null;
  agent_name: string | null;
  persona_name: string | null;
  rehearsee: string;
  channel: string;
  status: string;
  pending_action?: string | null;
  attempt_count?: number;
  retryable?: boolean;
  overall_score: number | null;
  objection_coverage: number | null;
  booking_attempted: boolean | null;
  tone_score: number | null;
  created_at: string;
  completed_at: string | null;
}

export interface RehearsalRun extends RehearsalRunSummary {
  max_turns: number;
  transcript: RehearsalTranscriptTurn[];
  scores: RehearsalScores;
  strengths: string[];
  gaps: string[];
  suggestions: string[];
  summary: string | null;
  error: string | null;
  updated_at: string;
}

export interface CreateRehearsalRequest {
  idempotency_key?: string;
  agent_id: string;
  persona_id: string;
  rehearsee?: RehearseeType;
  channel?: string | null;
  max_turns?: number;
}

export interface CreatePersonaRequest {
  name: string;
  description?: string;
  slug?: string;
  difficulty?: PersonaDifficulty;
  channel?: string;
  persona_prompt: string;
  opening_message?: string;
  objections?: string[];
  goal?: string;
}
